File: android/jvm-tests/src/test/kotlin/PairingTest.kt
import hu.breaker.app.core.Pairing
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Párosító kód — a `desktop/test/pairing.test.ts` párja.
 *
 * A KÓDNAK BÁJTRA EGYEZNIE KELL a gépen kiadottal. Ha nem, a felhasználó a
 * gépen kapott kódot begépeli a telefonba, és vagy nem történik semmi, vagy —
 * ami rosszabb — MÁS címre kerül a jelszava.
 *
 * A várt kódok itt nincsenek kiszámolva, csak BEMÁSOLVA: a TypeScript oldal
 * tényleges kimenete. Így a teszt nem azt bizonyítja, hogy a Kotlin kód
 * önmagával konzisztens, hanem hogy a kettő ugyanaz.
 */
class PairingTest {

    @Test
    fun `the code matches what the desktop produces, character for character`() {
        // Ezek nincsenek kiszámolva, csak BEMÁSOLVA: a TypeScript oldal
        // tényleges kimenetei. Így a teszt nem azt bizonyítja, hogy a Kotlin
        // kód önmagával konzisztens, hanem hogy a kettő UGYANAZ — enélkül a
        // gépen kiadott kód a telefonon nem nyílna ki.
        val fixtures = mapOf(
            "http://192.168.1.10:8787" to "00GMR",
            "http://10.0.0.5:8787" to "800019G",
            "http://172.20.3.4:8787" to "H030JR",
            "http://8.8.8.8:8787" to "R40G208N",
            "http://192.168.1.10:9000" to "4HJG08AM",
        )
        for ((url, want) in fixtures) {
            assertEquals(want, Pairing.encode(url), "eltérő kód: $url")
            assertEquals(url, Pairing.decode(want), "a gépen kiadott kód nem nyílik ki: $want")
        }
    }

    @Test
    fun `a home address becomes a very short code`() {
        // A rövidség a lényeg: ha a kód is tíz karakter, semmivel nem jobb az
        // IP-címnél.
        val code = Pairing.encode("http://192.168.1.10:8787")
        assertTrue(code != null && code.length == 5, "öt karakter, nem ${code?.length}")
        assertEquals("http://192.168.1.10:8787", Pairing.decode(code!!))
    }

    @Test
    fun `every private range survives the round trip`() {
        for (url in listOf(
            "http://192.168.0.1:8787",
            "http://192.168.255.255:8787",
            "http://10.0.0.5:8787",
            "http://10.255.255.254:8787",
            "http://172.16.4.9:8787",
            "http://172.31.200.1:8787",
            "http://100.64.3.7:8787",
            "http://8.8.8.8:8787",
            "http://192.168.1.10:9000",
        )) {
            val code = Pairing.encode(url)
            assertTrue(code != null, "nem kódolható: $url")
            assertEquals(url, Pairing.decode(code!!), "oda-vissza eltér: $url -> $code")
            assertTrue(code.length <= 8, "túl hosszú (${code.length}): $url")
        }
    }

    @Test
    fun `a mistyped code is refused, not silently pointed elsewhere`() {
        val good = Pairing.encode("http://192.168.1.10:8787")!!
        val alphabet = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
        var refused = 0
        var total = 0
        for (i in good.indices) {
            for (ch in alphabet) {
                if (ch == good[i]) continue
                total++
                val bad = good.substring(0, i) + ch + good.substring(i + 1)
                val decoded = Pairing.decode(bad)
                if (decoded == null) refused++
                else assertTrue(decoded != "http://192.168.1.10:8787")
            }
        }
        assertTrue(refused.toDouble() / total > 0.9, "csak ${refused * 100 / total}%-ot fogott meg")
    }

    @Test
    fun `hyphens, spaces and look-alikes never decide whether a code works`() {
        // Kézzel másolva senki nem figyel ezekre. Egy kód, ami emiatt nem megy,
        // ugyanolyan rossz, mint a be nem gépelt IP-cím.
        val code = Pairing.encode("http://192.168.1.10:8787")!!
        val want = "http://192.168.1.10:8787"
        assertEquals(want, Pairing.decode(code.lowercase()))
        assertEquals(want, Pairing.decode(Pairing.format(code)))
        assertEquals(want, Pairing.decode(" " + code.toCharArray().joinToString(" ") + " "))
        assertEquals(want, Pairing.decode(code.replace("1", "I").replace("0", "O")))
    }

    @Test
    fun `junk does not become a valid address`() {
        // A bizonytalanság az ELUTASÍTÁS felé dől: egy véletlenül elfogadott
        // kód olyan gépre küldené a jelszót, amiről a felhasználó nem tud.
        for (bad in listOf("", "   ", "ZZZZZ", "ABC", "nem egy kód", "!!!!!", "A".repeat(40))) {
            assertNull(Pairing.decode(bad), bad)
        }
        assertNull(Pairing.encode("https://192.168.1.10:8787"))
        assertNull(Pairing.encode("http://sync.pelda.hu:8787"))
        assertNull(Pairing.encode("nem egy cím"))
    }

    @Test
    fun `one field takes both a code and an address`() {
        val code = Pairing.encode("http://192.168.1.10:8787")!!
        assertEquals("http://192.168.1.10:8787", Pairing.resolveServerInput(code))
        assertEquals("http://192.168.1.10:8787", Pairing.resolveServerInput("http://192.168.1.10:8787"))
        assertEquals("https://sync.pelda.hu", Pairing.resolveServerInput("https://sync.pelda.hu"))
        assertEquals("http://192.168.1.10:8787", Pairing.resolveServerInput("192.168.1.10:8787"))
        assertNull(Pairing.resolveServerInput(""))
        assertNull(Pairing.resolveServerInput("   "))
    }
}

File: desktop/test/pairing.test.ts
// Párosító kód.
//
// Ez a funkció azon a ponton áll, ahol a szinkron eddig meghalt: be kellett
// gépelni egy IP-címet. Ha a kód rosszul működik, a felhasználó vagy nem tud
// belépni, vagy — ami rosszabb — MÁSHOVA lép be, mint hitte.
//
// Ezért a tesztek nem szép eseteket néznek, hanem a két kártékony kimenetelt:
// jó kódot elutasítani, és rossz kódból érvényes címet csinálni.

import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import {
  DEFAULT_SYNC_PORT, decodePairingCode, encodePairingCode, formatPairingCode,
  resolveServerInput,
} from '../src/shared/sync/pairing';

test('a home address becomes a very short code', () => {
  // A rövidség a lényeg: ha a kód is tíz karakter, semmivel nem jobb az
  // IP-címnél. Otthon szinte mindig 192.168.x.y van, és épp azt a két bájtot
  // kell leírni, ami változik.
  const code = encodePairingCode('http://192.168.1.10:8787');
  assert.ok(code, 'kódolhatónak kell lennie');
  assert.equal(code.length, 5, `öt karakter, nem ${code.length}`);
  assert.equal(decodePairingCode(code), 'http://192.168.1.10:8787');
});

test('every private range survives the round trip', () => {
  for (const url of [
    'http://192.168.0.1:8787',
    'http://192.168.255.255:8787',
    'http://10.0.0.5:8787',
    'http://10.255.255.254:8787',
    'http://172.16.4.9:8787',
    'http://172.31.200.1:8787',
    'http://100.64.3.7:8787',   // CGNAT — a „bármi” ág
    'http://8.8.8.8:8787',
  ]) {
    const code = encodePairingCode(url);
    assert.ok(code, `nem kódolható: ${url}`);
    assert.equal(decodePairingCode(code), url, `oda-vissza eltér: ${url} -> ${code}`);
    assert.ok(code.length <= 8, `túl hosszú (${code.length}): ${url}`);
  }
});

test('a non-default port is carried too', () => {
  const url = 'http://192.168.1.10:9000';
  const code = encodePairingCode(url)!;
  assert.equal(decodePairingCode(code), url);
  // Az alapértelmezett port NEM kerül a kódba — ezért rövidebb.
  assert.ok(code.length > encodePairingCode('http://192.168.1.10:8787')!.length);
});

test('a mistyped code is refused, not silently pointed elsewhere', () => {
  // Ez a legfontosabb teszt. Ellenőrző összeg nélkül egy elrontott betűből
  // MÁSIK, létező cím lenne, és a felhasználó csak annyit látna, hogy a
  // kiszolgáló nem érhető el — miközben valójában elgépelte a kódot.
  const good = encodePairingCode('http://192.168.1.10:8787')!;
  let refused = 0;
  let total = 0;
  const alphabet = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
  for (let i = 0; i < good.length; i++) {
    for (const ch of alphabet) {
      if (ch === good[i]) continue;
      total++;
      const bad = good.slice(0, i) + ch + good.slice(i + 1);
      const decoded = decodePairingCode(bad);
      if (decoded === null) refused++;
      else assert.notEqual(decoded, 'http://192.168.1.10:8787');
    }
  }
  // Öt bit ellenőrző összeg: az egykarakteres elgépelések nagy részét elkapja.
  assert.ok(refused / total > 0.9, `csak ${Math.round((refused / total) * 100)}%-ot fogott meg`);
});

test('hyphens, spaces and lower case never decide whether a code works', () => {
  // Kézzel másolva senki nem figyel ezekre. Egy kód, ami emiatt nem megy,
  // ugyanolyan rossz, mint a be nem gépelt IP-cím.
  const code = encodePairingCode('http://10.0.0.5:8787')!;
  const want = 'http://10.0.0.5:8787';
  assert.equal(decodePairingCode(code.toLowerCase()), want);
  assert.equal(decodePairingCode(formatPairingCode(code)), want);
  assert.equal(decodePairingCode(` ${code.split('').join(' ')} `), want);
});

test('look-alike characters land in the same place', () => {
  // A Crockford ábécében nincs I, L, O és U. Aki mégis annak látja az 1-est
  // vagy a 0-t, jó eredményt kapjon, ne hibaüzenetet.
  const code = encodePairingCode('http://192.168.1.10:8787')!;
  const messy = code.replace(/1/g, 'I').replace(/0/g, 'O');
  assert.equal(decodePairingCode(messy), 'http://192.168.1.10:8787');
});

test('junk does not become a valid address', () => {
  // A bizonytalanság itt az ELUTASÍTÁS felé dől: egy véletlenül elfogadott kód
  // olyan gépre küldené a jelszót, amiről a felhasználó nem tud.
  for (const bad of ['', '   ', 'ZZZZZ', 'ABC', 'nem egy kód', '!!!!!', 'A'.repeat(40)]) {
    assert.equal(decodePairingCode(bad), null, bad);
  }
  assert.equal(decodePairingCode(undefined as unknown as string), null);
});

test('what cannot be coded honestly returns null', () => {
  // Tartománynévre és HTTPS-re nem gyártunk kódot: az működésképtelen lenne,
  // és a felület inkább a teljes címet mutatja.
  assert.equal(encodePairingCode('https://192.168.1.10:8787'), null);
  assert.equal(encodePairingCode('http://sync.pelda.hu:8787'), null);
  assert.equal(encodePairingCode('nem egy cím'), null);
  assert.equal(encodePairingCode(undefined as unknown as string), null);
});

test('one field takes both a code and an address', () => {
  // Külön mező a kódnak és a címnek azt jelentené, hogy előbb el kell dönteni,
  // melyikbe kell írni — és pont az ilyen apró döntéseknél hagyják abba.
  const code = encodePairingCode('http://192.168.1.10:8787')!;
  assert.equal(resolveServerInput(code), 'http://192.168.1.10:8787');
  assert.equal(resolveServerInput('http://192.168.1.10:8787'), 'http://192.168.1.10:8787');
  assert.equal(resolveServerInput('https://sync.pelda.hu'), 'https://sync.pelda.hu');
  // Séma nélkül: a http:// hiánya nem hiba, csak kihagyott gépelés.
  assert.equal(resolveServerInput('192.168.1.10:8787'), 'http://192.168.1.10:8787');
  assert.equal(resolveServerInput('sync.pelda.hu'), 'http://sync.pelda.hu');
  assert.equal(resolveServerInput(''), null);
  assert.equal(resolveServerInput('   '), null);
});

test('the code reads back in groups, because it is copied by hand', () => {
  assert.equal(formatPairingCode('ABCDEFG'), 'ABCD-EFG');
  assert.equal(formatPairingCode('ABCD'), 'ABCD');
  assert.equal(DEFAULT_SYNC_PORT, 8787);
});
